onnxruntime/core/providers/dml/DmlExecutionProvider/src/AbiCustomRegistry.cpp

```diff
@@ -400,7 +400,15 @@ HRESULT STDMETHODCALLTYPE AbiCustomRegistry::RegisterOperatorKernel(
     {
         builder.InputMemoryType(::OrtMemType::OrtMemTypeCPUInput, 0);
     }
-        
+    else if (name == "Shape")
+    {
+        builder.OutputMemoryType(::OrtMemType::OrtMemTypeCPUInput, 0);
```

Comment:
>builder.OutputMemoryType(::OrtMemType::OrtMemTypeCPUInput, 0); [](http://example.com/codeflow?start=8&length=62)

One thing I wonder (maybe Jeff knows) is whether the Shape operator actually needs to wait for the GPU input tensor to get results, or whether it can work asynchronously when the shape is already known, since they don't care about the values in the tensor.

I'm glad you're adding this, but also we really shouldn't need to implement our own Shape or Size operator (nor should CUDA), because Shape and Size are technically EP agnostic given they only read the *description* of the tensor and not actually any of the data within the tensor. One implementation should work for all EP's :/.

Reply:
This is mostly a problem with our EP though because of how different it registers operators. Most other EPs just have a registration and straight up call the CPU implementation of Shape, but we have an ML wrapper over all that stuff.